### DISCUSSION TITLE
高版本(0.9.40-fix1之后)对基于musl的linux发行版不兼容

### DISCUSSION THREAD

**Discussion by hecheng337** [Mrs4s/go-cqhttp#1024] [Q&A] [answered]
事情是这样的：我今天上午突发奇想，尝试在我的软路由上运行bot。环境信息大致如下
![-34a4140ba3c5afae.png](https://user-images.githubusercontent.com/57560866/130731800-ecf2179c-fad2-427f-a168-3de689fa7fc3.png)
![30e66844ab2eab9d.png](https://user-images.githubusercontent.com/57560866/130731821-37d65782-099b-40c1-a2a0-006026e1e5f4.png)
总之，这玩意运行的是arm64架构上，使用了精简版(相较于libc)musl标准库的一个linux发行版(OpenWrt)
命令执行之后马上产生了奇怪的报错(图在下面...)。我想不对呀，去年底我的bot在路由器上挂了一个多星期来着。于是我一个一个tag往前翻，诶，还真是：
![40f7b1b246c18797_edit_11112384711845.png](https://user-images.githubusercontent.com/57560866/130732216-94d91df8-b25a-460b-a14c-921643973537.png)
从这个版本往前的，都可以完美运行；之后的，全部会出现类似报错(golang那个模块的版本号会变)
我没用过go语言，工具链长啥样都不知道。当时我就想着会不会是编译的时候有某个库没连接进去，然后这个库被OpenWrt删减了，于是我试着在路由器上配置了go环境，手动克隆mater分支编译了一遍。结果还是一样的：![)2%_0IJ(8_E1U1)E{9K]D0H.png](https://user-images.githubusercontent.com/57560866/130733670-e92dc1a6-0976-4c92-b218-9b24fb28beec.png)
正好我手头还有一台树莓派(arm架构，Ubuntu系统)，就顺便拿来对照测试了一下![Screenshot_20210825_131324_com.termux.png](https://user-images.githubusercontent.com/57560866/130733995-15bf363f-bd6a-41e8-9c61-0817ae008d30.png)
居然是可以正常运行最新版的。树莓派和我软路由用的是同款芯片，只是代数的区别。这样一来我能想到的二者唯一的区别就是：树莓派使用的是标准的glibc库，另一个则是精简版的musl(具体区别可以谷歌)。
其实我觉得在软路由上运行bot这个想法还挺...朴素的。毕竟软路由功耗低，不要钱(买了之后)，24h稳定在线，而且配置好插件之后一般是有完美的翻墙环境的。我在想如果可以的话，能不能在wiki的某个角落暗示一下后来者，别在路由器上折腾这玩意了，不要浪费时间😂或者有没有大佬能告诉我解决办法啥的(比如单独降golang模块版本...我实在是不熟，不会操作)确实可行的话，我也可以来帮忙写写wiki。如果有需要的话，欢迎联系我提供测试环境，开学之前我都很闲的XD再不济，在路由器上老一点的版本也不是不能用嘛🤨🤣
我待会自己再试试，有啥消息我第一时间更新过来。

**Comment by hecheng337** [comment]
[https://github.com/Mrs4s/go-cqhttp/commit/8dc2349a9b377456fbfdd5c3a2dfd66c131a40b6](url)
暂时定位到导致问题的gomod是在这个commit加入的

**Comment by Ink-33** [comment]
请问您在路由器上运行最新版（100b6）也是报同样的错误吗

**Reply by hecheng337** [reply]
是的。我最开始就是发现最新版用不了才会倒回去翻旧版的🤣
路由器上用的是从这个仓库
https://github.com/coolsnowwolf/lede
手动编译的固件(核心开发者叫lean大，在国内软路由圈子里非常有名，基本上大家用的OpenWrt都是用他这套工具编译生成的)，编译的时候能选的动态库我都编进去了，平时跑跑dotnet，java，node，python这些平台的应用也都没有问题。
我今天下午试了试，弄了一台x86的虚拟机跑软路由，也是一样的问题。另外，只要是在我上面说的那个commit之前的版本，就没有问题。基本上可以确定，只要是OpenWrt平台，就没法正常使用新版。
有软路由的朋友可以试试🧐

**Reply by Ink-33** [reply]
麻烦您使用ubuntu paste等工具贴一下您的`/etc/services`

**Reply by hecheng337** [reply]
贴在下面回复里了😬我看了看那个出错的go源码，好像确实是和/etc/service一样的作用，用来返回端口的信息啥的。我试试比较一下路由器上的这个文件和Ubuntu那些有啥区别

**Reply by hecheng337** [reply]
> 贴在下面回复里了😬我看了看那个出错的go源码，好像确实是和/etc/service一样的作用，用来返回端口的信息啥的。我试试比较一下路由器上的这个文件和Ubuntu那些有啥区别

草，不对劲，两边的文件差了远了。。。diff一下弹了几屏幕

**Reply by Ink-33** [reply]
![Screenshot_2021-08-25-22-38-18-658_com.sonelli.juicessh.jpg](https://user-images.githubusercontent.com/51873347/130811194-1f40b91a-d80b-4eca-8e57-b294eb07684d.jpg)

**Comment by Mrs4s** [comment]
应该是 `CCGO` 炸了. 可能有点麻烦

**Comment by Mrs4s** [comment]
能否发一下 `cat /etc/services` 的结果?

**Reply by hecheng337** [reply]
![Screenshot_20210825_222426_com.termux_edit_5964617170964.jpg](https://user-images.githubusercontent.com/57560866/130808837-76f0945f-4652-4889-b546-8778ea3aac88.jpg)

**Comment by wdvxdr1123** [comment, marked as answer]
临时解决方案: 拉取 dev 分支，使用 go build -tags nosilk 进行编译

**Reply by hecheng337** [reply]
![Screenshot_20210825_223844_com.termux.png](https://user-images.githubusercontent.com/57560866/130811192-689800e5-739a-4d42-9ce2-9006c2645015.png)
😉测试可行

**Comment by Mrs4s** [comment]
这是一个上游库的问题, 需要等待上游修复
